scheduler/scd_utils.py

```diff
@@ -0,0 +1,253 @@
+            ValueError: If line parameters are invalid or if line is a duplicate.
+        """
+
+        new_line = self.check_line(yyyymmdd, hhmm, prio, experiment, duration)
+
+        scd_lines = self.read_scd()
+
+        if new_line in scd_lines:
+            raise ValueError("Line is a duplicate of an existing line")
+
+        scd_lines.append(new_line)
+
+        # sort first by timestamp, then by priority, then by duration with default duration last.
+        # duration sorting is funky cause the default value is not a int.
+        new_scd = sorted(scd_lines, key=lambda x:(x['timestamp'],
+                                                    x['prio'],
+                                                    (x['duration'] == '-', x['duration'])))
+
+        self.write_scd(new_scd)
+
+    def remove_line(self, yyyymmdd, hhmm, prio, experiment, duration='-'):
+        """Summary
+
+        Args:
+            yyyymmdd (str): year/month/day string.
+            hhmm (str): hour/minute string.
+            prio (int or str): priority value.
+            experiment (str): The experiment to run.
+            duration (str, optional): an optional duration to run for.
+
+        Raises:
+            ValueError: If line parameters are invalid or if line does not exist.
+        """
+
+        line_to_rm = self.check_line(yyyymmdd, hhmm, prio, experiment, duration)
+
+        scd_lines = self.read_scd()
+        try:
+            scd_lines.remove(line_to_rm)
+        except:
+            raise ValueError("Line does not exist in SCD")
+
+        self.write_scd(scd_lines)
+
+    def get_relevant_lines(self, yyyymmdd, hhmm):
+        """Gets any relevant future lines given a supplied time. If a re
+
+        Args:
+            yyyymmdd (str): year/month/day string.
+            hhmm (str): hour/minute string.
+
+        Returns:
+            TYPE: List of relevant dicts of line info.
+
+        Raises:
+            ValueError: If datetime could not be created from supplied arguments.
+        """
+
+        try:
+            # create datetime from args to see if valid
+            time = dt.datetime.strptime(yyyymmdd + " " + hhmm, self.scd_dt_fmt)
+        except:
+            raise ValueError("Can not create datetime from supplied formats")
+
+        scd_lines = self.read_scd()
+
+        epoch = dt.datetime.utcfromtimestamp(0)
+        epoch_milliseconds = int((time - epoch).total_seconds() * 1000)
+
+        equals = False
+        prev_line_appended = False
+        relevant_lines = []
+        for idx, line in enumerate(scd_lines):
+            if line['timestamp'] == epoch_milliseconds:
+                equals = True
+                relevant_lines.append(line)
+            elif line['timestamp'] > epoch_milliseconds:
+                if equals:
+                    relevant_lines.append(line)
+                else:
+                    if not prev_line_appended:
+                        last_line_timestamp = scd_lines[idx-1]['timestamp']
```

Comment:
I might add a try... except IndexError here in the case that this index is the first in the schedule? If it is then prev_line_appended = True and continue

Reply:
alternatively just an if idx == 0 ... might be better

Reply:
Yeah I actually address this slightly differently in a later commit.